chain/index/ddls.go

```diff
@@ -0,0 +1,82 @@
+package index
+
+import "database/sql"
+
+const DefaultDbFilename = "chainindex.db"
+
+var ddls = []string{
+	`CREATE TABLE IF NOT EXISTS tipset_message (
+		message_id INTEGER PRIMARY KEY,
+		tipset_key_cid BLOB NOT NULL,
+		height INTEGER NOT NULL,
+		reverted INTEGER NOT NULL,
+		message_cid BLOB,
+		message_index INTEGER,
+		UNIQUE (tipset_key_cid, message_cid)
+	)`,
+
+	`CREATE TABLE IF NOT EXISTS eth_tx_hash (
+		tx_hash TEXT PRIMARY KEY,
+		message_cid BLOB NOT NULL,
+		inserted_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+	)`,
+
+	`CREATE TABLE IF NOT EXISTS event (
+		event_id INTEGER PRIMARY KEY,
+		message_id INTEGER NOT NULL,
+		event_index INTEGER NOT NULL,
+		emitter_addr BLOB NOT NULL,
+		reverted INTEGER NOT NULL,
```

Comment:
Can we not infer "reverted" from the message/chain tables? Or are we storing it here for performance?

Reply:
This is read performance. 

An event is reverted when the **_execution tipset_** for the corresponding message is reverted (even though the tipset that contains the message itself might still be un-reverted).

Putting this flag here avoids having to do one additional join for each `GetLogs` query -> which really is one of the most heavily used ETH RPCs.

Reply:
It is not a premature optimisation. We've already run into performance problems with reads for the events API for archival nodes.

Every additional join penalises read performance. Rodd recently noticed a difference of 10% in perf because of an additional join.

Reply:
We've run into performance issues, yes. But those are generally not related to the databases unless I'm mistaken. Archival performance issues usually stem from the blockstores.

> Rodd recently noticed a difference of 10% in perf just because of an additional join.

It depends on the join. But again, the question is: does it actually matter here. I would be very surprised if normalizing these tables (assuming all the correct indices) is going to be the bottleneck for indexing.

Reply:
For the record, when doing https://github.com/filecoin-project/lotus/pull/12571, I thought I could optimise the AMT building by doing a single query for all of the event_entries for a single tipset instead of one query per message; then in memory just sort them out locally (very easy operation, especailly if they are sorted out of the db). But that involved a `event_entry JOIN event` so I could select on the `event` table instead. But that resulted in a ~10% slowdown across a range of tipsets (with many events) instead of a speed-up from doing one vs many queries!

But yeah, it's going to depend on the join and the indexes involved. It did leave me wondering about what overhead SQLite gets involved when you invoke a `JOIN`, I suspect there's a whole lot of additional mechanics that hit you in the performance-bone that aren't present if you're just fishing around in a separate table.

Reply:
@Stebalien 

-> Say we don't add a reverted field here.
-> An event is reverted when it's "execution tipset" is rerverted (even though the corresponding message inclusion tipset might not be reverted).
-> So, even if we remove the `reverted` flag here -> we then need to replace it with a "execution_tipset" field here and then do a join with the tipsets table to find out if the "execution_tipset" has been reverted or not.
-> So, not only do we still need to keep around an extra field -> we also need to do an additional join to get the same info.
-> "reverted" here really is a field that tracks "the execution tipset has been reverted".

Reply:
So yeah sorry -> I think my original comment wasn't accurate. This is NOT for de-normalization. This is to keep track of message execution.

Reply:
> An event is reverted when it's "execution tipset" is rerverted (even though the corresponding message inclusion tipset might not be reverted).

No? The message inclusion tipset is the execution tipset. When I ask "what events happened in tipset X", I expect to get the events associated with the messages in tipset X.

Yes, we _happen_ to execute tipsets when we receive a new block that references the tipset as its parent, but that doesn't mean reverting that block reverts the tipset execution.

Reply:
@Stebalien Not sure what you mean here.

 The **_existing_** event Index reverts events when tipset **_execution_** is reverted, even if tipset inclusion is NOT reverted.

See  https://github.com/filecoin-project/lotus/blob/d4b8af99b201053a7ae40397052fe75522c867ac/chain/events/filter/index.go#L288 AND https://github.com/filecoin-project/lotus/blob/d4b8af99b201053a7ae40397052fe75522c867ac/chain/events/filter/event.go#L364

`revert` is passed as true when the execution tipset is reverted and the events in the inclusion tipset are then marked as reverted which makes sense to me.

Let's discuss this sync today.